File: java/pl/edu/pg/mygame/organisms/animals/Antelope.java
package pl.edu.pg.mygame.organisms.animals;

import pl.edu.pg.mygame.Grid;
import pl.edu.pg.mygame.World;
import pl.edu.pg.mygame.organisms.Organism;
import pl.edu.pg.mygame.organisms.Species;

import java.awt.*;
import java.util.List;

import static pl.edu.pg.mygame.utils.HelperFunctions.getRandomPoint;
import static pl.edu.pg.mygame.utils.HelperFunctions.isLucky;

public class Antelope extends Animal {
    public Antelope(Point p) {
        super(p);
        this.species = Species.ANTELOPE;
        this.initiative = 4;
        this.strength = 4;
    }

    @Override
    protected Point getNewPoint(Grid grid) {
        List<Point> far_neighbours = grid.getDeepNeighbourCells(position);
        // when there are no cells at range 2, get the cells at range 1
        // this can for example happen when square grid is 3x3 and antelope is in the middle
        if (far_neighbours.size() == 0)
            far_neighbours = grid.getNeighbourCells(position);
        return getRandomPoint(far_neighbours);
    }

    @Override
    public boolean collision(Organism other, boolean attacking, Grid grid) {
        if (isLucky(50)) {
            List<Point> points = grid.getFreeNeighbourCells(position);
            if (points.size() == 0)
                return false;
            Point p = getRandomPoint(points);
            Point old_position = position;
            grid.setOrganism(position, null);
            position = p;
            grid.setOrganism(position, this);
            World.message_panel.showMessage("Antelope escaped from the fight to:" +
                    String.format("(%d, %d)\n", position.x, position.y));

            if (!attacking) {
                grid.setOrganism(other.getPosition(), null);
                other.setPosition(old_position);
                grid.setOrganism(old_position, other);
            }
            return true;
        }
        return false;
    }
}

File: python/organisms/animals/antelope.py
from gui.logger import Logger
from organisms.animals.animal import Animal
from organisms.species import Species
from utils.helpers import Helpers


class Antelope(Animal):
    def __init__(self, pos):
        super().__init__(pos)
        self.strength = 4
        self.initiative = 4
        self.species = Species.ANTELOPE

    def get_new_point(self, grid):
        far_neighbors = grid.get_deep_neighbor_cells(self.position)
        # when there are no cells at range 2, get the cells at range 1
        # this can for example happen when square grid is 3x3 and antelope is in the middle
        if len(far_neighbors) == 0:
            far_neighbors = grid.get_neighbor_cells(self.position)
        return Helpers.get_random_point(far_neighbors)

    def collision(self, other, attacking, grid):
        if Helpers.is_lucky(50):
            points = grid.get_free_neighbor_cells(self.position)
            if len(points) == 0:
                return False
            p = Helpers.get_random_point(points)
            old_position = self.position
            grid.set_organism(self.position, None)
            self.position = p
            grid.set_organism(self.position, self)
            Logger.log(f'Antelope escaped from the fight to: {self.position}')

            if not attacking:
                grid.set_organism(other.position, None)
                other.position = old_position
                grid.set_organism(old_position, other)
            return True
        return False
